src/core/parser.y

```diff
@@ -73,7 +73,7 @@ int lexerlex_destroy(void);
 int lexerlex(void);
 static void handle_assignment(const std::string token, Expression *expr, const Location loc);
 
-std::stack<LocalScope *> scope_stack;
+std::stack<std::shared_ptr<LocalScope>> scope_stack;
```

Comment:
When we do the final pop() in `parse()`, I wonder if we should assert that this stack is empty, just in case a scope somehow leaks.